svm/src/account_loader.rs

```diff
@@ -404,93 +449,51 @@ fn load_transaction_accounts<CB: TransactionProcessingCallback>(
 }
 
 fn load_transaction_account<CB: TransactionProcessingCallback>(
-    callbacks: &CB,
+    account_loader: &mut AccountLoader<CB>,
     message: &impl SVMMessage,
     account_key: &Pubkey,
     account_index: usize,
-    instruction_accounts: &[&u8],
-    account_overrides: Option<&AccountOverrides>,
     feature_set: &FeatureSet,
     rent_collector: &dyn SVMRentCollector,
-    loaded_programs: &ProgramCacheForTxBatch,
 ) -> Result<(LoadedTransactionAccount, bool)> {
     let mut account_found = true;
-    let is_instruction_account = u8::try_from(account_index)
-        .map(|i| instruction_accounts.contains(&&i))
-        .unwrap_or(false);
-    let is_writable = message.is_writable(account_index);
+    let usage_pattern = AccountUsagePattern::new(message, account_index);
+
     let loaded_account = if solana_sdk::sysvar::instructions::check_id(account_key) {
         // Since the instructions sysvar is constructed by the SVM and modified
-        // for each transaction instruction, it cannot be overridden.
+        // for each transaction instruction, it cannot be loaded.
         LoadedTransactionAccount {
             loaded_size: 0,
             account: construct_instructions_account(message),
             rent_collected: 0,
         }
-    } else if let Some(account_override) =
-        account_overrides.and_then(|overrides| overrides.get(account_key))
+    } else if let Some(mut loaded_account) = account_loader.load_account(account_key, usage_pattern)
     {
+        loaded_account.rent_collected = if usage_pattern == AccountUsagePattern::Writable {
```

Comment:
This change has the side effect of collecting rent from writable accounts that come from account overrides. Probably fine? I'm not too sure yet

Reply:
this was intentional:
* the current code actually collects rent from an override if its the fee-payer, otherwise it doesnt. so this is at least consistent
* with simd83, we have to consider what happens to a writable overridden account after the first transaction: does it stay frozen, and all changes to it are ignored? or does it behave like an ordinary account that happens to have an overridden starting value? the latter is more in line with its current behavior, where it is freely allowed to change in the transaction, and then the mutated value is retained in the `LoadedTransactionAccount` returned from the svm entrypoint
* if we let an overriden account change intrabatch then rent collection should happen on first use, otherwise we collect on _second_ use, which is odd
* account overrides is only intended to be used for simulation. actually giving it a writable account for a real execution would be highly unwise because it would make its way all the way back to accounts-db to be committed

in fact, account overrides is an overly general mechanism for transaction simulation to pass svm the previous slot view of the `SlotHistory` sysvar and nothing else. andrew was suggesting we could change it if/when we change the `load_and_execute_sanitized_transactions` interface

Reply:
Great breakdown, completely onboard now